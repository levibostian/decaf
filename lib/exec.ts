import { DeployCommandOutput, isDeployCommandOutput } from "./steps/types/deploy.ts"
import * as log from "./log.ts"
import * as shellQuote from "shell-quote"
import { DeployEnvironment } from "./types/environment.ts"

export interface RunResult {
  exitCode: number
  stdout: string
  output: DeployCommandOutput | undefined
}

export interface Exec {
  run: (
    { command, input, displayLogs }: {
      command: string
      input: DeployEnvironment | undefined
      displayLogs?: boolean
      envVars?: { [key: string]: string }
      throwOnNonZeroExitCode?: boolean
    },
  ) => Promise<RunResult>
}

/*
Executes a command and returns the exit code and the stdout of the command.

The entire command is passed in as 1 string. This is for convenience but also because when used as a github action, the commands will be passed to the tool as a single string.
Then, the command is split into the command and the arguments. This is required by Deno.Command.
You cannot simply split the string by spaces to create the args list. Example, if you're given: `python3 -c "import os; print(os.getenv('INPUT'))"` we expect only 2 args: "-c" and "import ...".
We use a popular package to parse the string into the correct args list. See automated tests to verify that this works as expected.

To make this function testable, we not only have the stdout and stderr be piped to the console, but we return it from this function so tests can verify the output of the command.
*/
const run = async (
  { command, input, displayLogs, envVars, throwOnNonZeroExitCode }: {
    command: string
    input: DeployEnvironment | undefined
    displayLogs?: boolean
    envVars?: { [key: string]: string }
    throwOnNonZeroExitCode?: boolean
  },
): Promise<
  { exitCode: number; stdout: string; output: DeployCommandOutput | undefined }
> => {
  // If command actually contains 2 commands (using &&), throw an error. The API of this function simply doesn't support that.
  if (command.includes("&&")) {
    throw new Error(
      `The command "${command}" contains multiple commands (uses &&). This is not supported. Please run each command separately.`,
    )
  }

  if (displayLogs) {
    log.message(` $> ${command}`)
  } else {
    log.debug(` $> ${command}`)
  }

  const execCommand = command.split(" ")[0]
  const execArgs = shellQuote.parse(
    command.replace(new RegExp(`^${execCommand}\\s*`), ""),
  )
  const environmentVariablesToPassToCommand: { [key: string]: string } = envVars || {}

  // For some features to work, we need to communicate with the command. We need to send data to it and read data that it produces.
  // We use JSON as the data format to communicate with the command since pretty much every language has built-in support for it.
  // Since we are creating subprocesses to run the command, we are limited in how we can communicate with the command.
  // One common way would be to ask the subprocess to stdout a JSON string that we simply read, but this tool tries to promote stdout
  // as a way to communicate with the user, not the tool. So instead, we write the JSON to a file and pass the file path to the command.
  let tempFilePathToCommunicateWithCommand: string | undefined
  let inputDataFileContents: string | undefined
  if (input) {
    tempFilePathToCommunicateWithCommand = await Deno.makeTempFile({
      prefix: "new-deployment-tool-",
      suffix: ".json",
    })
    inputDataFileContents = JSON.stringify(input)
    await Deno.writeTextFile(
      tempFilePathToCommunicateWithCommand,
      inputDataFileContents,
    )

    environmentVariablesToPassToCommand["DATA_FILE_PATH"] = tempFilePathToCommunicateWithCommand
  }

  // We want to capture the stdout of the command but we also want to stream it to the console. By using streams, this allows us to
  // output the stdout/stderr to the console in real-time instead of waiting for the command to finish before we see the output.
  const process = new Deno.Command(execCommand, {
    args: execArgs,
    stdout: "piped",
    stderr: "piped",
    env: environmentVariablesToPassToCommand,
  })

  const child = process.spawn()

  let capturedStdout = ""
  let capturedStderr = ""

  child.stdout.pipeTo(
    new WritableStream({
      write(chunk) {
        const decodedChunk = new TextDecoder().decode(chunk)

        if (displayLogs) {
          log.message(decodedChunk)
        } else {
          log.debug(decodedChunk)
        }

        capturedStdout += decodedChunk.trimEnd()
      },
    }),
  )
  child.stderr.pipeTo(
    new WritableStream({
      write(chunk) {
        const decodedChunk = new TextDecoder().decode(chunk)

        if (displayLogs) {
          log.message(decodedChunk)
        } else {
          log.debug(decodedChunk)
        }

        capturedStderr += decodedChunk.trimEnd()
      },
    }),
  )

  const code = (await child.status).code
  if (capturedStdout) log.debug(capturedStdout)
  if (capturedStderr) log.debug(capturedStderr)

  let commandOutput: DeployCommandOutput | undefined

  if (tempFilePathToCommunicateWithCommand) {
    const outputDataFileContents = await Deno.readTextFile(
      tempFilePathToCommunicateWithCommand,
    )
    const commandOutputUntyped = JSON.parse(outputDataFileContents)
    // As long as the command wrote something to the file and the type is correct, we will use it.
    if (
      isDeployCommandOutput(commandOutputUntyped) &&
      outputDataFileContents !== inputDataFileContents
    ) { // there is a chance that the command did not write to the file or they have a bug.
      commandOutput = commandOutputUntyped
    }
  }

  log.debug(
    `exit code, ${code}, command output: ${JSON.stringify(commandOutput)}`,
  )

  let shouldThrowError = true
  if (throwOnNonZeroExitCode !== undefined && throwOnNonZeroExitCode == false) {
    shouldThrowError = false
  }

  if (code !== 0 && shouldThrowError) {
    throw new Error(`Command: ${command}, failed with exit code: ${code}, output: ${capturedStdout}, stderr: ${capturedStderr}`)
  }

  return {
    exitCode: code,
    stdout: capturedStdout,
    output: commandOutput,
  }
}

export const exec: Exec = {
  run,
}
