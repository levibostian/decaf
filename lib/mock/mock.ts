import { stub, spy, GetParametersFromProp, GetReturnFromProp, Stub } from "jsr:@std/testing@1/mock";

/**
 * Example: 
 * ```
 * interface GitHubActions {
 *  isRunningInPullRequest(): Promise<boolean>
 * }
 * 
 * const githubActions: GitHubActions = mock()
 * when(githubActions, "isRunningInPullRequest", async () => true)
 * 
 * // optionally, get information about the mock: 
 * const isRunningInPullRequestMock = when(...)
 * assertEquals(isRunningInPullRequestMock.calls.length, 1)
 * ```
 */

export function when<T, Prop extends keyof T>(
  mockObject: T, 
  functionName: Prop, 
  thenCall: (this: T, ...args: GetParametersFromProp<T, Prop>) => GetReturnFromProp<T, Prop>): Stub<T, GetParametersFromProp<T, Prop>, GetReturnFromProp<T, Prop>> {
    return stub(mockObject, functionName, thenCall)
}

export function mock<T>(): T {
  const mockObject = {} as T;
  
  // for convenience, stub all functions with an empty implementation. So if a mock function doesn't need any logic, it can be used without needing to call `when` on it.
  for (const key in mockObject) {
    mockObject[key] = stub(mockObject, key) as unknown as T[Extract<keyof T, string>];
  }
  
  return mockObject;
}
