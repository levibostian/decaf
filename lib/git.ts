import { exec } from "./exec.ts"
import { Exec } from "./exec.ts"
import { GitHubCommit } from "./github-api.ts"
import * as log from "./log.ts"

export interface Git {
  checkoutBranch: (
    { exec, branch, createBranchIfNotExist }: { exec: Exec; branch: string; createBranchIfNotExist: boolean },
  ) => Promise<void>
  merge: (
    { exec, branchToMergeIn, commitTitle, commitMessage, fastForward }: {
      exec: Exec
      branchToMergeIn: string
      commitTitle: string
      commitMessage: string
      fastForward?: "--no-ff" | "--ff-only"
    },
  ) => Promise<void>
  pull: ({ exec }: { exec: Exec }) => Promise<void>
  setUser: (
    { exec, name, email }: { exec: Exec; name: string; email: string },
  ) => Promise<void>
  squash: (
    { exec, branchToSquash, branchMergingInto, commitTitle, commitMessage }: {
      exec: Exec
      branchToSquash: string
      branchMergingInto: string
      commitTitle: string
      commitMessage: string
    },
  ) => Promise<void>
  rebase: (
    { exec, branchToRebaseOnto }: { exec: Exec; branchToRebaseOnto: string },
  ) => Promise<void>
  getLatestCommitsSince({ exec, commit }: { exec: Exec; commit: GitHubCommit }): Promise<GitHubCommit[]>
  getLatestCommitOnBranch({ exec, branch }: { exec: Exec; branch: string }): Promise<GitHubCommit>
  createLocalBranchFromRemote: ({ exec, branch }: { exec: Exec; branch: string }) => Promise<void>
  getCommits: ({ exec, branch }: { exec: Exec; branch: string }) => Promise<GitHubCommit[]>
}

const checkoutBranch = async (
  { exec, branch, createBranchIfNotExist }: { exec: Exec; branch: string; createBranchIfNotExist: boolean },
): Promise<void> => {
  await exec.run({
    command: `git checkout ${createBranchIfNotExist ? "-b " : ""}${branch}`,
    input: undefined,
  })
}

const merge = async (
  { exec, branchToMergeIn, commitTitle, commitMessage, fastForward }: {
    exec: Exec
    branchToMergeIn: string
    commitTitle: string
    commitMessage: string
    fastForward?: "--no-ff" | "--ff-only"
  },
): Promise<void> => {
  await exec.run({
    command: `git merge ${branchToMergeIn} -m "${commitTitle}" -m "${commitMessage}" ${fastForward || ""}`,
    input: undefined,
  })
}

const pull = async ({ exec }: { exec: Exec }): Promise<void> => {
  await exec.run({
    command: `git pull`,
    input: undefined,
  })
}

const setUser = async (
  { exec, name, email }: { exec: Exec; name: string; email: string },
): Promise<void> => {
  await exec.run({
    command: `git config user.name "${name}"`,
    input: undefined,
  })

  await exec.run({
    command: `git config user.email "${email}"`,
    input: undefined,
  })
}

// Squash all commits of a branch into 1 commit
const squash = async (
  { exec, branchToSquash, branchMergingInto, commitTitle, commitMessage }: {
    exec: Exec
    branchToSquash: string
    branchMergingInto: string
    commitTitle: string
    commitMessage: string
  },
): Promise<void> => {
  // We need to find out how many commits 1 branch is ahead of the other to find out how many unique commits there are.
  const { stdout } = await exec.run({
    command: `git rev-list --count ${branchMergingInto}..${branchToSquash}`,
    input: undefined,
  })

  const numberOfCommitsAheadOfBranchMergingInto = parseInt(stdout.trim())

  if (numberOfCommitsAheadOfBranchMergingInto === 0) {
    log.message(`Branches ${branchToSquash} and ${branchMergingInto} are already up to date. No commits to squash.`)
    return
  }

  // Now that we know how many commits are ahead, we can squash all of those commits into 1 commit.
  await exec.run({
    command: `git reset --soft HEAD~${numberOfCommitsAheadOfBranchMergingInto}`,
    input: undefined,
  })
  await exec.run({
    command: `git commit -m "${commitTitle}" -m "${commitMessage}"`,
    input: undefined,
  })
}

const rebase = async (
  { exec, branchToRebaseOnto }: { exec: Exec; branchToRebaseOnto: string },
): Promise<void> => {
  await exec.run({
    command: `git rebase ${branchToRebaseOnto}`,
    input: undefined,
  })
}

const getLatestCommitsSince = async (
  { exec, commit }: { exec: Exec; commit: GitHubCommit },
): Promise<GitHubCommit[]> => {
  const { stdout } = await exec.run({
    command: `git log --pretty=format:"%H|%s|%ci" ${commit.sha}..HEAD`,
    input: undefined,
  })

  return stdout.trim().split("\n").map((commitString) => {
    const [sha, message, dateString] = commitString.split("|")

    return { sha, message, date: new Date(dateString) }
  })
}

const getLatestCommitOnBranch = async (
  { exec, branch }: { exec: Exec; branch: string },
): Promise<GitHubCommit> => {
  const { stdout } = await exec.run({
    command: `git log -1 --pretty=format:"%H|%s|%ci" ${branch}`,
    input: undefined,
  })

  const [sha, message, dateString] = stdout.trim().split("|")

  return { sha, message, date: new Date(dateString) }
}

/**
 * Makes sure that we have a local branch that has all of the commits that the remote branch has.
 *
 * There are a lot of commands here, just to get a local branch of a remote branch. After many attempts to simplify this, we would hit many different errors.
 * I believe that complexity comes because we run this tool on a CI server where the git config might be different.
 * Running all of these commands and running each command by itself (example: not running `git checkout -b` to try and combine creating a branch and checking it out)
 * have given the most consistent results.
 */
const createLocalBranchFromRemote = async (
  { exec, branch }: { exec: Exec; branch: string },
): Promise<void> => {
  const currentBranchName = (await exec.run({
    command: `git branch --show-current`,
    input: undefined,
  })).stdout.trim()
  const doesBranchExist = (await exec.run({
    command: `git branch --list ${branch}`,
    input: undefined,
  })).stdout.trim() !== ""

  // Perform a fetch, otherwise you might get errors about origin branch not being found.
  await exec.run({
    command: `git fetch origin`,
    input: undefined,
  })

  // Only run if it doesn't exist locally. This is to avoid a error that crashes the tool: "fatal: a branch named '<branch-name>' already exists"
  if (!doesBranchExist) {
    await exec.run({
      command: `git branch --track ${branch} origin/${branch}`,
      input: undefined,
    })
  }

  // Checkout the branch so we can pull it.
  await exec.run({
    command: `git checkout ${branch}`,
    input: undefined,
  })

  // Pull the branch from the remote.
  // Adding --no-rebase to avoid an error that could happen when you run pull.
  // The error is: You have divergent branches and need to specify how to reconcile them.
  await exec.run({
    command: `git pull --no-rebase origin ${branch}`,
    input: undefined,
  })

  // Switch back to the branch we were on before.
  await exec.run({
    command: `git checkout ${currentBranchName}`,
    input: undefined,
  })
}

const getCommits = async (
  { exec, branch }: { exec: Exec; branch: string },
): Promise<GitHubCommit[]> => {
  const currentBranchName = (await exec.run({
    command: `git branch --show-current`,
    input: undefined,
  })).stdout.trim()

  await checkoutBranch({ exec, branch, createBranchIfNotExist: false })

  const { stdout } = await exec.run({
    command: `git log --pretty=format:"%H|%s|%ci"`,
    input: undefined,
  })

  const commits = stdout.trim().split("\n").map((commitString) => {
    const [sha, message, dateString] = commitString.split("|")

    return { sha, message, date: new Date(dateString) }
  })

  await checkoutBranch({ exec, branch: currentBranchName, createBranchIfNotExist: false })

  return commits
}

export const git: Git = {
  checkoutBranch,
  merge,
  pull,
  setUser,
  squash,
  rebase,
  getCommits,
  getLatestCommitsSince,
  getLatestCommitOnBranch,
  createLocalBranchFromRemote,
}
