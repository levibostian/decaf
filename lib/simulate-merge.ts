import { Exec } from "./exec.ts"
import { Git } from "./git.ts"
import { GitHubCommit } from "./github-api.ts"

export interface SimulateMerge {
  performSimulation(
    type: "merge" | "rebase" | "squash",
    { baseBranch, targetBranch, commitTitle, commitMessage }: {
      baseBranch: string
      targetBranch: string
      commitTitle: string
      commitMessage: string
    },
  ): Promise<GitHubCommit[]>
  merge: (
    { baseBranch, targetBranch, commitTitle, commitMessage }: {
      baseBranch: string
      targetBranch: string
      commitTitle: string
      commitMessage: string
    },
  ) => Promise<GitHubCommit[]>
  squash: (
    { baseBranch, targetBranch, commitTitle, commitMessage }: {
      baseBranch: string
      targetBranch: string
      commitTitle: string
      commitMessage: string
    },
  ) => Promise<GitHubCommit[]>
  rebase: (
    { baseBranch, targetBranch, commitTitle, commitMessage }: {
      baseBranch: string
      targetBranch: string
      commitTitle: string
      commitMessage: string
    },
  ) => Promise<GitHubCommit[]>
}

export class SimulateMergeImpl implements SimulateMerge {
  constructor(private git: Git, private exec: Exec) {}

  performSimulation(
    simulateMergeType: "merge" | "rebase" | "squash",
    { baseBranch, targetBranch, commitTitle, commitMessage }: {
      baseBranch: string
      targetBranch: string
      commitTitle: string
      commitMessage: string
    },
  ) {
    switch (simulateMergeType) {
      case "merge":
        return this.merge({ baseBranch, targetBranch, commitTitle, commitMessage })
      case "rebase":
        return this.rebase({ baseBranch, targetBranch, commitTitle, commitMessage })
      case "squash":
        return this.squash({ baseBranch, targetBranch, commitTitle, commitMessage })
    }
  }

  // Perform a merge, including creating a merge commit
  async merge(
    { baseBranch, targetBranch, commitTitle, commitMessage }: {
      baseBranch: string
      targetBranch: string
      commitTitle: string
      commitMessage: string
    },
  ) {
    const commitReference = await this.git.getLatestCommitOnBranch({ exec: this.exec, branch: targetBranch })

    await this.prepareForMerge({ baseBranch, targetBranch })

    await this.git.checkoutBranch({ exec: this.exec, branch: targetBranch, createBranchIfNotExist: false })

    await this.git.merge({ exec: this.exec, branchToMergeIn: baseBranch, commitTitle, commitMessage })

    return await this.git.getLatestCommitsSince({ exec: this.exec, commit: commitReference })
  }

  async squash(
    { baseBranch, targetBranch, commitTitle, commitMessage }: {
      baseBranch: string
      targetBranch: string
      commitTitle: string
      commitMessage: string
    },
  ) {
    const commitReference = await this.git.getLatestCommitOnBranch({ exec: this.exec, branch: targetBranch })

    await this.prepareForMerge({ baseBranch, targetBranch })

    await this.git.checkoutBranch({ exec: this.exec, branch: baseBranch, createBranchIfNotExist: false })

    // perform a rebase to make sure that the squash we perform will only have commits for this branch.
    // if baseBranch, for example, contains a merge commit, if we do not rebase, that merge commit's changes will be included in the squash.
    // this may revert changes that the target branch has made.
    await this.git.rebase({ exec: this.exec, branchToRebaseOnto: targetBranch })

    // Squash all commits in PR into 1 commit.
    await this.git.squash({ exec: this.exec, branchToSquash: baseBranch, branchMergingInto: targetBranch, commitTitle, commitMessage })

    // we want to merge the squashed commit into the target branch.
    await this.git.checkoutBranch({ exec: this.exec, branch: targetBranch, createBranchIfNotExist: false })
    await this.git.merge({ exec: this.exec, branchToMergeIn: baseBranch, commitTitle, commitMessage, fastForwardOnly: true })

    return await this.git.getLatestCommitsSince({ exec: this.exec, commit: commitReference })
  }

  // Perform a rebase without creating a merge commit
  async rebase(
    { baseBranch, targetBranch, commitTitle, commitMessage }: {
      baseBranch: string
      targetBranch: string
      commitTitle: string
      commitMessage: string
    },
  ) {
    const commitReference = await this.git.getLatestCommitOnBranch({ exec: this.exec, branch: targetBranch })

    await this.prepareForMerge({ baseBranch, targetBranch })

    await this.git.checkoutBranch({ exec: this.exec, branch: baseBranch, createBranchIfNotExist: false })
    await this.git.rebase({ exec: this.exec, branchToRebaseOnto: targetBranch })
    await this.git.checkoutBranch({ exec: this.exec, branch: targetBranch, createBranchIfNotExist: false })
    await this.git.merge({ exec: this.exec, branchToMergeIn: baseBranch, commitTitle, commitMessage, fastForwardOnly: true })

    return await this.git.getLatestCommitsSince({ exec: this.exec, commit: commitReference })
  }

  private async prepareForMerge({ baseBranch, targetBranch }: { baseBranch: string; targetBranch: string }) {
    // First, make sure we have the latest changes from both branches
    await this.git.checkoutBranch({ exec: this.exec, branch: baseBranch, createBranchIfNotExist: false })
    await this.git.pull({ exec: this.exec })
    await this.git.checkoutBranch({ exec: this.exec, branch: targetBranch, createBranchIfNotExist: false })
    await this.git.pull({ exec: this.exec })

    // Setup git to be able to create a merge commit
    // The values here do not matter because we are not pushing the changes back to the remote
    await this.git.setUser({ exec: this.exec, name: "Deployment Test", email: "test@test.com" })
  }
}
