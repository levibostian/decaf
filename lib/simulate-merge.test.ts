import { assertEquals, assertFalse, assertRejects } from "@std/assert"
import { afterEach, before, beforeEach, describe, it } from "@std/testing/bdd"
import { assertSpyCall, restore, Stub, stub } from "@std/testing/mock"
import { exec, RunResult } from "./exec.ts"
import { SimulateMerge, SimulateMergeImpl } from "./simulate-merge.ts"
import { git } from "./git.ts"
import { Exec } from "./exec.ts"
import { assertSnapshot } from "@std/testing/snapshot"

describe("snapshot test all of the merge options", () => {
  let simulateMerge: SimulateMerge
  let execMock: Stub<Exec>

  beforeEach(() => {
    execMock = stub(exec, "run", async (args) => {
      // rev-list expects to return a number of commits
      if (args.command.includes("git rev-list")) {
        return { exitCode: 0, stdout: "3", output: undefined }
      }

      // log expects to return a list of commits
      // I am not the biggest fan of having to do this (I would prefer to mock the git module functions directly), but because we need to capture the git commands being executed, we have to put these mocks here.
      if (args.command.includes("git log")) {
        return {
          exitCode: 0,
          stdout:
            `[[⬛]]f15d5ac43a47b8333019461170fcaf0bd5a139d4[⬛]test: update snapshots to remove un-used ones[⬛]test: update snapshots to remove un-used ones
[⬛]Foo Bar[⬛]foo@bar.com[⬛]Foo Bar[⬛]foo@bar.com[⬛]2025-07-09 07:02:35 -0500[⬛]2c7e53b35e1b4b278700294b7781d770f16124c8[⬛]
0       42      __snapshots__/deploy.test.ts.snap`,
          output: undefined,
        }
      }

      return { exitCode: 0, stdout: "success", output: undefined }
    })
    simulateMerge = new SimulateMergeImpl(git, exec)
  })

  afterEach(() => {
    restore()
  })

  it("merge, should generate the correct git commands, should end on target branch", async (t) => {
    await simulateMerge.performSimulation("merge", {
      baseBranch: "feature",
      targetBranch: "main",
      pullRequestNumber: 123,
      pullRequestTitle: "title-here",
      pullRequestDescription: "message-here",
    })

    await assertSnapshot(t, execMock.calls.map((call) => call.args[0].command))

    assertEndsOnTargetBranch("main")
  })

  it("squash, should generate the correct git commands, should end on target branch", async (t) => {
    await simulateMerge.performSimulation("squash", {
      baseBranch: "feature",
      targetBranch: "main",
      pullRequestNumber: 123,
      pullRequestTitle: "title-here",
      pullRequestDescription: "message-here",
    })

    await assertSnapshot(t, execMock.calls.map((call) => call.args[0].command))

    assertEndsOnTargetBranch("main")
  })

  it("rebase, should generate the correct git commands, should end on target branch", async (t) => {
    await simulateMerge.performSimulation("rebase", {
      baseBranch: "feature",
      targetBranch: "main",
      pullRequestNumber: 123,
      pullRequestTitle: "title-here",
      pullRequestDescription: "message-here",
    })

    await assertSnapshot(t, execMock.calls.map((call) => call.args[0].command))

    assertEndsOnTargetBranch("main")
  })

  const assertEndsOnTargetBranch = (targetBranch: string) => {
    const lastCheckoutCommand = execMock.calls.filter((call) => call.args[0].command.includes("git checkout")).pop()
    assertEquals(lastCheckoutCommand?.args[0].command.includes(targetBranch), true)
  }
})
