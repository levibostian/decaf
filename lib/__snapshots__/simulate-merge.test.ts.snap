export const snapshot = {};

snapshot[`snapshot test all of the merge options > merge, should generate the correct git commands, should end on target branch 1`] = `
[
  'git log -1 --pretty=format:"%H|%s|%ci" main',
  "git checkout feature",
  "git branch --show-current",
  "git pull origin success",
  "git checkout main",
  "git branch --show-current",
  "git pull origin success",
  'git config user.name "Deployment Test"',
  'git config user.email "test@test.com"',
  "git checkout main",
  'git merge feature -m "Merge pull request #123 from feature" -m "" --no-ff',
  'git log --pretty=format:"%H|%s|%ci" success..HEAD',
]
`;

snapshot[`snapshot test all of the merge options > squash, should generate the correct git commands, should end on target branch 1`] = `
[
  'git log -1 --pretty=format:"%H|%s|%ci" main',
  "git checkout feature",
  "git branch --show-current",
  "git pull origin success",
  "git checkout main",
  "git branch --show-current",
  "git pull origin success",
  'git config user.name "Deployment Test"',
  'git config user.email "test@test.com"',
  "git checkout feature",
  "git rebase main",
  "git rev-list --count main..feature",
  "git reset --soft HEAD~3",
  'git commit -m "title-here" -m "message-here"',
  "git checkout main",
  'git merge feature -m "title-here" -m "message-here" --ff-only',
  'git log --pretty=format:"%H|%s|%ci" success..HEAD',
]
`;

snapshot[`snapshot test all of the merge options > rebase, should generate the correct git commands, should end on target branch 1`] = `
[
  'git log -1 --pretty=format:"%H|%s|%ci" main',
  "git checkout feature",
  "git branch --show-current",
  "git pull origin success",
  "git checkout main",
  "git branch --show-current",
  "git pull origin success",
  'git config user.name "Deployment Test"',
  'git config user.email "test@test.com"',
  "git checkout feature",
  "git rebase main",
  "git checkout main",
  'git merge feature -m "title-here" -m "message-here" --ff-only',
  'git log --pretty=format:"%H|%s|%ci" success..HEAD',
]
`;
